src/node/waveshaper.rs

```diff
@@ -369,11 +371,11 @@ impl WaveShaperRenderer {
         let wave_in = input.channels();
 
         let up_wave_in = self.upsampler_x2.process(wave_in).unwrap();
-        let mut up_wave_out = up_wave_in.clone();
+        let mut up_wave_out = up_wave_in.clone(); // this looks like allocation
```

Comment:
You are correct. We're actually not using the cloned values anyway. We can probably insert a reusable buffer here

Reply:
or maybe even mutate `up_wave_in ` in place.

I will try to have a look